src/commands/watch.rs

```diff
@@ -145,116 +173,126 @@ impl WatchCommand {
+                };
+                let sources = build
+                    .workdir
+                    .clone()
+                    .map(|workdir| {
+                        watch
+                            .iter()
+                            .filter_map(|w| workdir.join(w).to_str().map(String::from))
+                            .collect()
+                    })
+                    .unwrap_or(watch);
+                Some(
+                    sources
+                        .into_iter()
+                        .map(|s| WatchPattern::new(s, !self.skip_build, app_dir.as_path()))
+                        .collect::<Vec<Result<WatchPattern>>>(),
+                )
             })
-            .flatten();
-        let component_source_patterns = app_manifest
+            .flatten()
+            .collect::<Result<Vec<WatchPattern>>>()?;
+
+        // We want to watch a component's source if it has no build step. If we're skipping
+        // building then we'll watch all component sources.
+        let mut artifact_patterns = app_manifest
             .components
             .iter()
             .filter_map(|c| {
-                if let RawModuleSource::FileReference(path) = &c.source {
-                    return path.to_str();
-                }
-                None
+                let RawModuleSource::FileReference(path) = &c.source else {
+                    return None;
+                };
+                let path_str = path.to_str()?;
+                let watch_this_source = self.skip_build || c.build.is_none();
+                Some(WatchPattern::new(
+                    path_str.to_owned(),
+                    watch_this_source,
+                    app_dir.as_path(),
+                ))
             })
-            .map(String::from);
-        let component_file_patterns = app_manifest
-            .components
-            .iter()
-            .filter_map(|c| c.wasm.files.as_ref())
-            .flatten()
-            .filter_map(|raw_file_mount| match raw_file_mount {
-                RawFileMount::Placement(raw_directory_placement) => raw_directory_placement
-                    .source
-                    .join("**/*")
-                    .to_str()
-                    .map(String::from),
-                RawFileMount::Pattern(pattern) => Some(pattern.to_string()),
-            });
-
-        let mut path_patterns: Vec<String> = match self.skip_build {
-            false => watch_patterns.chain(component_file_patterns).collect(),
-            true => component_source_patterns
-                .chain(component_file_patterns)
-                .collect(),
-        };
-        // Always watch spin.toml
-        path_patterns.push(String::from(
-            self.app.to_str().expect("conversion to str not to fail"),
-        ));
+            .collect::<Result<Vec<WatchPattern>>>()?;
 
-        Ok(path_patterns)
+        // We always want to watch component files
+        artifact_patterns.append(
```

Comment:
Consider giving the files iterator a name so you can write something like `let artifact_patterns = concat(source_patterns, files_patterns)`.